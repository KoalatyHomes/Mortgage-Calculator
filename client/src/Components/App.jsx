import React, { useState } from 'react';
import PreQualified from './PreQualified.jsx';

function App() {
  const [homePriceVal, sethomePriceVal] = useState(4798000);
  const [downPaymentRate, setPaymentRate] = useState(20);
  const [setDownPaymentTotal] = useState(750);
  const [interestRate, setInterestRate] = useState(3.49);
  const [loanType, setLoanType] = useState('30-year fixed');

  let mortgageIns = 0;

  const nf = new Intl.NumberFormat(); // adds commas
  const downPaymentTotal = Math.round((homePriceVal * downPaymentRate) / 100);
  const homePriceFormat = `$${nf.format(homePriceVal)}`;
  const downPaymentFormat = `$${nf.format(downPaymentTotal)}`;
  const downPaymentRateFormat = `${downPaymentRate}%`;
  const interestRateFormat = `${interestRate}%`;

  /* *** COMPOUND INTEREST FORMULA *** */
  const months = 15;
  const N = months * 12;
  const P = homePriceVal - downPaymentTotal;
  const r = Number(((interestRate / 100).toFixed(5) / 12).toFixed(5));
  const top = ((1 + r) ** N).toFixed(3);
  const principalInterestVal = Math.ceil(P * ((r * top) / ((top) - 1)));

  /* *** PROPERTY TAX FORMULA *** */
  const propertyTaxes = Math.ceil((homePriceVal * 0.01) / 12);

  /* *** HOME INSURANCE *** */
  const insurance = 75;

  /* *** MORTAGE INS. FORMULA *** */
  if ((downPaymentTotal) === 0) {
    mortgageIns = 0;
  } else if ((downPaymentRate / 100) < 0.20) {
    mortgageIns = Math.ceil(100 / (downPaymentRate / 100));
  } else {
    mortgageIns = 0;
  }

  /* *** MONTHLY MORTGAGE PAYMENTS *** */
  const monthlyMortgage = nf.format(principalInterestVal + propertyTaxes + insurance + mortgageIns);

  return (
    <div className="pageLayout">
      <div className="affordability-container">
        <div className="affordability-text">
          Affordability
        </div>
        <div className="calculation-container">
          <div className="calculation-text">
            Calculate your monthly mortgage payments
          </div>
          <div className="est-payment">
            Your est. payment: $
            {monthlyMortgage}
            /month
          </div>
        </div>
        <div className="input-container">
          <div className="flex-container-padding">
            <div className="grid-container-padding">
              {/* **************** HOME PRICE **************** */}
              <div className="grid-cell-box grid-flex">
                <div className="input-container ">
                  <div className="financial-detail ">
                    <div className="text-container">
                      Home Price
                    </div>
                    <input id="homePriceInput" className="home-price-input" style={{ width: '112px' }} type="text" onChange={(e) => sethomePriceVal(e.target.value)} value={homePriceFormat} />
                  </div>
                  <div className="slider-container">
                    {/* Home Price Slider */}
                    <input className="range" type="range" min="0" max="8000000" step="10" name="homePriceVal" onChange={(e) => sethomePriceVal(e.target.value)} value={homePriceVal} />
                  </div>
                </div>
              </div>

              {/* ************** DOWN PAYMENT ****************** */}
              <div className="grid-cell-box grid-flex">
                <div className="input-container ">
                  <div className="financial-detail ">
                    <div className="text-container">
                      {/* <label forhtml="DownPaymentInput">Down Payment</label> */}
                      Down Payment
                    </div>
                    <div className="down-payment-wrapper">
                      {/* left payment */}
                      <input className="left-split-financial-input" id="DownPaymentInput" style={{ width: '100px' }} type="text" onChange={(e) => setDownPaymentTotal(e.target.value)} value={downPaymentFormat} />
                      {/* Down Payment right Textarea */}
                      <input className="right-split-financial-input" style={{ width: '56px' }} type="text" onChange={(e) => setPaymentRate(e.target.value)} value={(downPaymentRateFormat)} />
                    </div>
                  </div>
                  <div className="slider-container">
                    {/* Down Paypment Slider */}
                    <input className="range" type="range" min="0" max="30" onChange={(e) => setPaymentRate(e.target.value)} value={downPaymentRate} />
                  </div>
                </div>
              </div>

              {/* ************** INTEREST RATE ****************** */}
              <div className="grid-cell-box grid-flex">
                <div className="input-container ">
                  <div className="financial-detail ">
                    <div className="text-container">
                      {/* <label forhtml="InterestRateInput">Interest Rate</label> */}
                      Interest Rate
                    </div>
                    <input id="InterestRateInput" className="interest-rate-input" style={{ width: '75px' }} type="text" onChange={(e) => setInterestRate(e.target.value)} value={interestRateFormat} />
                  </div>
                  <div className="slider-container">
                    <input className="range" type="range" min="0.0" max="6.50" step="0.10" onChange={(e) => setInterestRate(e.target.value)} value={interestRate} />
                  </div>
                </div>
              </div>

              {/* ************** LOAN TYPE****************** */}
              <div className="grid-cell-box grid-flex">
                <div className="input-container ">
                  <div className="financial-detail ">
                    <div className="text-container">
                      {/* <label forhtml="LoanTypeInput">Loan Type</label> */}
                      Loan Type
                    </div>
                  </div>
                  <div className="select-container">
                    <div className="drop-down-container">
                      <div className="drop-down-arrow ">
                        {loanType}
                      </div>
                      <div className="svg-container">
                        <svg className="svg" viewBox="0 0 32 32" xmlns="http://www.w3.org/2000/svg">
                          <path d="M15.961 18.183l7.056-7.147 1.893 1.868-8.951 9.068-8.927-9.069 1.896-1.866z" fill="#869099"> </path>
                        </svg>
                      </div>
                    </div>
                    <select id="LoanTypeInput" value={loanType} className="select-input" onChange={(e) => { setInterestRate(Number((e.target.value).slice(0, 4))); setLoanType((e.target.value).slice(5)); }}>
                      <option value="3.49 30-year fixed">30-year fixed</option>
                      <option value="3.27 20-year fixed">20-year fixed</option>
                      <option value="2.86 15-year fixed">15-year fixed</option>
                      <option value="2.97 10-year fixed">10-year fixed</option>
                      <option value="3.06 FHA 30-year fixed">FHA 30-year fixed</option>
                      <option value="2.89 FHA 15-year fixed">FHA 15-year fixed</option>
                      <option value="2.84 VA 30-year fixed">VA 30-year fixed</option>
                      <option value="2.69 VA 15-year fixed">VA 15-year fixed</option>
                    </select>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      {/* ************** GRAPH CONTAINER  ****************** */}
      <div className="graph-container">
        {/* ************** VISUAL GRAPH  ****************** */}
        <div className="affordability-donut-graph">
          { /* https://medium.com/@heyoka/scratch-made-svg-donut-pie-charts-in-html5-2c587e935d72 */}
          <svg width="100%" height="100%" viewBox="0 0 42 42" className="donut">
            <circle className="donut-hole" cx="21" cy="21" r="15.91549430918954" fill="#fff"> </circle>
            <circle className="donut-ring" cx="21" cy="21" r="15.91549430918954" fill="transparent" stroke="#d2d3d4" strokeWidth="3"> </circle>
            <circle className="donut-segment" cx="21" cy="21" r="15.91549430918954" fill="transparent" stroke="#ce4b99" strokeWidth="3"> </circle>
          </svg>
        </div>
        {/* ************** GRAPH INFO ****************** */}
        <div className="affordability-table">
          {/* **************** PRINCIPLE AND INTEREST **************** */}
          <div className="payment-container seperate-cols">
            <div className="payment-cell">
              <div className="payment-content">
                <div className="payment-cell" style={{ width: '100%' }}>
                  <div className="payment-title-dot principal-dot">{/**/}</div>
                  <div className="payment-content">
                    Principal & Interest
                  </div>
                </div>
              </div>
              <div className="price-text">
                $
                {nf.format(principalInterestVal)}
              </div>
            </div>
          </div>
          {/* **************** PROPERTY TAX **************** */}
          <div className="payment-container seperate-cols">
            <div className="payment-cell">
              <div className="payment-content">
                <div className="payment-cell" style={{ width: '100%' }}>
                  <div className="payment-title-dot property-tax-dot">{/* to get rid of syntax error */}</div>
                  <div className="payment-content">
                    Property Taxes
                  </div>
                </div>
              </div>
              <div className="price-text">
                $
                {nf.format(propertyTaxes)}
              </div>
            </div>
          </div>
          {/* **************** HOME INSURANCE **************** */}
          <div className="payment-container seperate-cols">
            <div className="payment-cell">
              <div className="payment-content">
                <div className="payment-cell" style={{ width: '100%' }}>
                  <div className="payment-title-dot home-insurance-dot"> </div>
                  <div className="payment-content">
                    Home Insurance
                  </div>
                </div>
              </div>
              <div className="price-text">
                $
                {insurance}
              </div>
            </div>
          </div>
          {/* **************** MORTGAGE INS. & OTHER **************** */}
          <div className="payment-container seperate-cols">
            <div className="payment-cell">
              <div className="payment-content">
                <div className="payment-cell" style={{ width: '100%' }}>
                  <div className="payment-title-dot mortgage-ins-dot"> </div>
                  <div className="payment-content">
                    Mortgage ins. & other
                  </div>
                </div>
              </div>
              <div className="price-text">
                $
                {nf.format(mortgageIns)}
              </div>
            </div>
          </div>
          {/* TO DO: increase padding between cells later */}
          <div className="extra-padding"> </div>
          <div>
            <PreQualified />
          </div>
        </div>
      </div>
    </div>

  );
}

export default App;
