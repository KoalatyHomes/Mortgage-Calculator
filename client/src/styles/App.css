html, body {
  font-family: sans-serif;
  color: rgb(59, 65, 68);
  background-color: white;
  font-size: 16px;
  line-height: 0.5px;
  letter-spacing: 0.8px;
}


/*layout for slider*/
input[type="range"]{
  -webkit-appearance: none;
  background: linear-gradient(to right, rgb(0, 120, 130) 0%, rgb(0, 120, 130) 50%, rgb(205, 209, 212) 50%, rgb(205, 209, 212) 100%);;
  width: 100%;
  height: 2px;
  margin-bottom: 2px; /* fixes input field overlapping range slider during resize*/
}

input[type="range"]::-webkit-slider-thumb {
  -webkit-appearance: none;
  background-color: teal;
  width: 14px;
  height: 14px;
  border-radius: 7px;
  /* adds the outer border on the slider thumb*/
  -webkit-box-shadow: 0px 0px 0px 5px white;
  -moz-box-shadow: 0px 0px 0px 5px white;
  box-shadow: 0px 1px 2px 5px white;
}

a{
  text-decoration: none;
  color: teal;
  font-size: 14px;
}

.pageLayout {
  max-width: 940px;
  margin: auto;
}

.affordabilityContainer {
  display: flex;
  flex-direction: column;
  margin: 32px 0px;
}

.graphContainer {
  display: flex;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  justify-content: center;
}

.affordabilityTable {
  box-sizing: border-box;
  display: flex;
  justify-content: center;
  flex-direction: column;
  min-width: 50%; /*  prevent payment type and amount from over lapping*/
  flex-grow: 1;
  flex-shrink: 1;
  flex-basis: 0%;
  padding: 16px 0px 16px 16px;
}

.extraPadding{
  padding: 0px 0px 32px;
}