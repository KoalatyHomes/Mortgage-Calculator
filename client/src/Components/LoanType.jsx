import React from 'react';
// import PropTypes from 'prop-types';
import styles from '../styles/FinancialDetails.css';

const LoanType = (props) => {
  const {
    loanType, setInterestRate, setLoanType,
  } = props;

  const handleChange = (e) => {
    const {value } = e.target;
    console.log('value', value);
    setInterestRate(Number((e.target.value).slice(0, 4)));
    setLoanType(e.target.value.slice(5));
  };

  return (
    <div className={styles.gridCellBox && styles.gridFlex}>
      <div className={styles.inputContainer}>
        <div className={styles.financialDetail}>
          <div className={styles.textContainer}>

            Loan Type
          </div>
        </div>

        <div className={styles.selectContainer}>
          <div className={styles.dropDownContainer}>
            <div className={styles.dropDownArrow}>
              {loanType}
            </div>
            <div className={styles.svgContainer}>
              <svg className={styles.svg} viewBox="0 0 32 32" xmlns="http://www.w3.org/2000/svg">
                <path d="M15.961 18.183l7.056-7.147 1.893 1.868-8.951 9.068-8.927-9.069 1.896-1.866z" fill="#869099"> </path>
              </svg>
            </div>
          </div>

          <select id="LoanTypeInput" value={loanType} className={styles.selectInput} onChange={handleChange}>
            <option value="3.49 30-year fixed">30-year fixed</option>
            <option value="3.27 20-year fixed">20-year fixed</option>
            <option value="2.86 15-year fixed">15-year fixed</option>
            <option value="2.97 10-year fixed">10-year fixed</option>
            <option value="3.06 FHA 30-year fixed">FHA 30-year fixed</option>
            <option value="2.89 FHA 15-year fixed">FHA 15-year fixed</option>
            <option value="2.84 VA 30-year fixed">VA 30-year fixed</option>
            <option value="2.69 VA 15-year fixed">VA 15-year fixed</option>
          </select>
        </div>
      </div>
    </div>
  );
};

// LoanType.propTypes = {
//   loanType: PropTypes.string.isRequired,
//   setInterestRate: PropTypes.func.isRequired,
//   setLoanType: PropTypes.func.isRequired,
// };

export default LoanType;
