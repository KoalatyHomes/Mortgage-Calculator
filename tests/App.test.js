/* eslint-disable react/jsx-filename-extension */
/* eslint-disable padded-blocks */
/* eslint-disable no-undef */
/* eslint-disable no-unused-vars */

import React from 'react';
import { shallow, mount } from 'enzyme';
import { render, fireEvent, screen, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import toJson from 'enzyme-to-json';
import { renderHook, act } from '@testing-library/react-hooks';

import App from '../client/src/Components/App';
import LoanType from '../client/src/Components/LoanType';

global.fetch = require('node-fetch');

afterEach(cleanup);

describe('App component', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it('renders correctly enzyme', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    expect(wrapper.exists()).toBe(true);
  });

});
