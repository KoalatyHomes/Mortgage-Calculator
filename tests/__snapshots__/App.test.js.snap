// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders correctly enzyme 1`] = `
<div>
  <div>
    <CostDetails
      monthlyMortgage="31,640"
    />
    <FinancialDetails
      downPaymentRate={20}
      downPaymentTotal={963800}
      homePriceVal={4819000}
      interestRate={3.49}
      loanType="30-year fixed"
      setDownPaymentTotal={963800}
      setInterestRate={[Function]}
      setLoanType={[Function]}
      setPaymentRate={[Function]}
      sethomePriceVal={[Function]}
    />
  </div>
  <div>
    <DonutGraph
      homeIns={75}
      monthlyMortgage={31640}
      mortgageIns={0}
      principal={27549}
      propertyTaxes={4016}
    />
    <div>
      <PaymentContainer
        homeIns={75}
        mortgage={0}
        principal={27549}
        taxes={4016}
      />
      <div>
         
      </div>
      <PreQualified />
    </div>
  </div>
</div>
`;
